Validation/RecoTau/plugins/TrackFromPVSelector.cc

```diff
@@ -1,100 +1,78 @@
+};
 
 
 ////////////////////////////////////////////////////////////////////////////////
 // construction/destruction
 ////////////////////////////////////////////////////////////////////////////////
 
 //______________________________________________________________________________
-TrackFromPVSelector::TrackFromPVSelector(const edm::ParameterSet& iConfig)
-  : max_dxy_          ( iConfig.getParameter<double>( "max_dxy" ) )
-  , max_dz_           ( iConfig.getParameter<double>( "max_dz" ) )
-  , v_recoVertexToken_( consumes< std::vector<reco::Vertex> >( iConfig.getParameter<edm::InputTag>( "srcVertex" ) ) )
-  , v_recoTrackToken_ ( consumes< std::vector<reco::Track> >( iConfig.getParameter<edm::InputTag>( "srcTrack" ) ) )
+TrackFromPVSelector::TrackFromPVSelector(edm::ParameterSet const& iConfig)
+  : max_dxy_{iConfig.getParameter<double>("max_dxy")}
+  , max_dz_{iConfig.getParameter<double>("max_dz")}
+  , v_recoVertexToken_{consumes<std::vector<reco::Vertex>>(iConfig.getParameter<edm::InputTag>("srcVertex"))}
+  , v_recoTrackToken_{consumes<std::vector<reco::Track>>(iConfig.getParameter<edm::InputTag>("srcTrack"))}
 {
-  produces<std::vector<reco::Track> >();
+  produces<std::vector<reco::Track>>();
 }
 
-
-//______________________________________________________________________________
-TrackFromPVSelector::~TrackFromPVSelector(){}
-
 ////////////////////////////////////////////////////////////////////////////////
 // implementation of member functions
 ////////////////////////////////////////////////////////////////////////////////
 
 //______________________________________________________________________________
-void TrackFromPVSelector::produce(edm::Event& iEvent,const edm::EventSetup& iSetup)
-{  
-  std::unique_ptr<std::vector<reco::Track> > goodTracks(new std::vector<reco::Track >);
-  
-  edm::Handle< std::vector<reco::Vertex> > VertexHandle;
-  iEvent.getByToken( v_recoVertexToken_, VertexHandle );
-
-  edm::Handle< std::vector<reco::Track> > TrackHandle;
-  iEvent.getByToken( v_recoTrackToken_, TrackHandle );
-  
-  if( (VertexHandle->size() == 0) || (TrackHandle->size() == 0) ) 
-  {
-    iEvent.put(std::move(goodTracks));
-    return ;
-  }
-  
-  reco::Vertex PV = VertexHandle->front();   
-  //typename std::vector<reco::Track>::const_iterator TrackIt ;
-  std::vector<reco::Track>::const_iterator TrackIt ;
-
-  for (TrackIt = TrackHandle->begin(); TrackIt != TrackHandle->end(); ++TrackIt) {
-    if ( fabs(TrackIt->dxy(PV.position())) < max_dxy_ &&
-         fabs(TrackIt->dz(PV.position()))  < max_dz_  ){
-      goodTracks -> push_back(*TrackIt) ;
+void TrackFromPVSelector::produce(edm::StreamID, edm::Event& iEvent, edm::EventSetup const&) const
+{
+  auto goodTracks = std::make_unique<std::vector<reco::Track>>();
+
+  edm::Handle<std::vector<reco::Vertex>> vertices;
+  edm::Handle<std::vector<reco::Track>> tracks;
+  if (iEvent.getByToken(v_recoVertexToken_, vertices) && !vertices->empty() &&
```

Comment:
This is not the same logic. Previously if the data product was missing we'd throw an exception at the `VertexHandle->` or `TrackHandle->` call. Now you will put in an empty container. Please remove the check on return value of `getByToken`.